### ISSUE TITLE
Masters should not be excluded from service load balancers

### ISSUE THREAD

**Issue by ljani** [kubernetes/kubernetes#65618]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
I'm running a single node cluster on AWS EC2, ie. scheduling work on the master as well. Now that I tried to add an ELB for my service, the EC2 instance is not associated with the ELB. The ELB gets created but there are 0 EC2 instances associated with it. I'm using ELB for SSL termination, if you wonder why I'd like to load balance a single node cluster.

The service controller logs this message:

```
I0628 17:54:48.853175       1 event.go:221] Event(v1.ObjectReference{Kind:"Service", Namespace:"default", Name:"nginx", UID:"59980933-7afc-11e8-9a8c-0621b993d602", APIVersion:"v1", ResourceVersion:"2052", FieldPath:""}): type: 'Warning' reason: 'UnAvailableLoadBalancer' There are no available nodes for LoadBalancer service default/nginx
```

**What you expected to happen**:
The EC2 instance is associated with the ELB.

**How to reproduce it (as minimally and precisely as possible)**:
- Boot an EC2 instance with correct IAM permissions
- Install a single node cluster using `kubeadm` and `cloud-provider=aws`
- Untaint the node
- Schedule eg. nginx
- Create an external load balancer for nginx
- View the ELB in AWS console and note the `Status: 0 of 0 instances in service`

**Anything else we need to know?**:
The reason for this seems to be the `node-role.kubernetes.io/master` label. [This](https://github.com/kubernetes/kubernetes/commit/82d600bb5c34b2c7352643d4d811941e063a8215#diff-56c9c66bf151b9d7268939da6ebf3613R609) blocks associating a load balancer with the node. On the other hand [this](https://github.com/kubernetes/kubernetes/commit/976e046f7a8d220dfd1df2e2f61d4630bcfbaf4c) changed what is included, because `includeNodeFromNodeList` did not check if a node is a master. I'm not sure what would be correct fix. I could try submit a PR, if you guide me how this should behave. Is my scenario even a supported one?

I think this bug should be reproducible on other clouds as well.

**Environment**:
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:08:34Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release):

```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

- Kernel (e.g. `uname -a`):

```
Linux ip-x.eu-central-1.compute.internal 4.9.0-6-amd64 #1 SMP Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64 GNU/Linux
```

- Install tools: `kubeadm`

**Comment by jhorwit2** [comment]
@ljani This is by design, load balancers do not include master nodes in the available pool of backend servers. You should remove the `node-role.kubernetes.io/master` label from your node and ensure it's schedulable & ready in order for it to be a backend. (exact filter logic lives [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L585-L619))

I'm going to close this issue as it's not a bug. If you would like to see the ability to use masters as available backends for LB's then please create another ticket with a feature request. 

/close

**Comment by ljani** [comment]
@jhorwit2 Thanks for the response. So, it's okay to remove the `node-role.kubernetes.io/master` label even if there's only one node in the cluster? Are there any drawbacks? I was a bit wary of that, because I thought each cluster would need at least one master and you should only remove the `NoSchedule` taint from it. The [kubeadm guide](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/) only refers to removing the taint from the master and not the label itself, but yeah, the guide is for a *single master* cluster, so it would be a little contradictory.

Searching for masterless does not really yield any related results.

Anyhow, if that's a supported scenario, then I'm very happy with it. Otherwise I should open the ticket.

**Comment by jhorwit2** [comment]
It all depends on how the cluster is setup and if they depend on the labels or not. Single node clusters will probably always have some quirks.

**Comment by ljani** [comment]
I've been following the [Creating a single master cluster with kubeadm](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/) guide and thus I'm using `kubeadm`. I'll try and see how it goes.

**Comment by jhorwit2** [comment]
The docs are probably incorrect when it comes to this because it wasn't tested. 

/reopen 
/remove-sig node
/sig cluster-lifecycle

**Comment by neolit123** [comment]
please, mind that kubeadm documentation with desist from adding cloud provider specific documentation in the future, but instead link to external resources.

/cc @kubernetes/sig-cluster-lifecycle-bugs 
/cc @kubernetes/sig-cloud-provider-bugs

**Comment by smarterclayton** [comment]
My comment in the other issue:

> I'm not sure this is correct as we run more things on the master. If I have a service that I want to expose like an API proxy, an aggregated API server that wants its own LB, or a not quite control plane but not quite end user workload. We shouldn't bias towards the master (I.e. a lot of node port services for regular workloads should bypass the master), but this PR as it stands prevents using service load balancers with services that run on masters, which limits options like self hosting and use.

> Now that node load balancing is more nuanced (with health checking endpoints), do we even need this? Masters should not be marked as healthy when pods aren't run on them, which means no traffic would go to masters?

I think excluding masters from service load balancer is a bug and needs a more nuanced design.  I'm going to bump priority here because I've got users trying to run pods on masters that have service LB on them that can't use service LB.

**Comment by smarterclayton** [comment]
Changing title to be more accurate.

**Comment by smarterclayton** [comment]
I think the correct fix here was the service LB health check support for targeting the serving pool to nodes that hold the pod, and now that we have that we don't need this hack.

**Comment by andrewsykim** [comment]
To be clear, this would only apply for Services with `externalTrafficPolicy: Local` right? Is it okay to assume that users will always use "Local" for this case?

**Comment by andrewsykim** [comment]
I'm sure we _could_ add logic in service controller to include master if externalTrafficPolicy == Local, exclude otherwise but that doesn't seem like an elegant solution

**Comment by smarterclayton** [comment]
For masters, it seems likely.  I mean, in general I expect externalTrafficPolicy: Local to be the correct setting for the vast majority of LB services - is there a reason I would *not* want that policy in general use?

**Comment by andrewsykim** [comment]
I general I would agree that LBs would use `externalTrafficPolicy: Local`. One example where I wouldn't' though is if I have a service that needs to ingress traffic both from an LB and from another service in the same cluster. In that case, I would consider using `externalTrafficPolicy: Cluster` in which case I may not want the master to be in the pool of LB backends. Though to be fair this may a rare case not worth looking into.

**Comment by jhorwit2** [comment]
@smarterclayton this should perhaps be merged with https://github.com/kubernetes/kubernetes/issues/65013

The service controller today does some filtering based on masters in interesting ways. Historically it seems that masters were marked unschedulable to be excluded from lb services. Then once labels for roles became popular that got added as well.

**Comment by jhorwit2** [comment]
> I'm sure we could add logic in service controller to include master if externalTrafficPolicy == Local, exclude otherwise but that doesn't seem like an elegant solution thinking

This gets a little hacky with how backends are updated today. We would need to keep track of two different backend sets while updating each service.

**Comment by andrewsykim** [comment]
> One example where I wouldn't' though is if I have a service that needs to ingress traffic both from an LB and from another service in the same cluster.

^ I was wrong about this because I had assumed cluster IP only routes to local too, but just realized it does cluster-wide regardless.

**Comment by smarterclayton** [comment]
As we increasingly work to split out core, there are more components running on masters (controllers, aggregate apis, web hook servers, admin debug services, etc).  A blanket "don't include masters" for a generic kube feature isn't appropriate, but I do agree that in general it is a reasonable default for some users on the cluster.  The important thing is we figure out what the right way to split out "non-control plane components and users" from "being able to use the infrastructure".  Everyone is free to install and run things on their masters, unless you're an opinionated cloud provider.

**Comment by andrewsykim** [comment]
This is fair, given most LBs use `externalTrafficPolicy: Local` I think adding masters by default and opting out with an annotation (or similar) would be a good approach.

**Comment by jhorwit2** [comment]
>  The important thing is we figure out what the right way to split out "non-control plane components and users" from "being able to use the infrastructure". Everyone is free to install and run things on their masters, unless you're an opinionated cloud provider.

Agreed, to me it sounds like the `ServiceNodeExclusion` feature currently in alpha could help facilitate this by having admins add labels to nodes to exclude them from receiving traffic and remove the "is master" style checks.

**Comment by andrewsykim** [comment]
It's too bad we can't exclude nodes on a per-service basis (yet). But seems like a reasonable compromise for now. Didn't even realize `ServiceNodeExclusion` was a feature, in which case I think we should PR something for adding masters back into the backend pool. @jhorwit2 do you have bandwidth to work on this?

**Comment by justinsb** [comment]
Other users did consider traffic going through nodes they had marked as a master to be a bug, so we should tread carefully.

That said, we may be able to come up with a design where `externalTrafficPolicy: Local` on a service can trump the node annotations.  If I add tolerations so that a pod can be scheduled on a node tainted as a master, and I have a service that targets that pod that is `externalTrafficPolicy: Local`, then that seems like a clear statement of intent that I _do_ want traffic to target that master node.

I'd echo what @thockin said - there are trade-offs here, so I'd caution against implementing anything you're not happy to throw away, before getting consensus on design (that we've ideally sanity-checked with people that considered it a bug the other way round to OP!)

**Comment by andrewsykim** [comment]
sgtm! I'll add this to the next SIG CP meeting agenda which is ~ 2 weeks from now. Hopefully that gives us enough time to think this over and discuss with interested parties involved.

**Comment by smarterclayton** [comment]
If users don't want nodes to be part of load balancers, we should have a separate annotation / label that is used to exclude nodes from load balancer consideration (which common install paths could set).  Keying off masters is much too specific a change.

**Comment by jhorwit2** [comment]
We have that label today via the service exclusion feature gate. So if we did remove the logic around masters users would have a feature in place today (we'd probably want to promote it past alpha) to remove them from service lbs.

**Comment by andrewsykim** [comment]
Had a discussion about this in our SIG meeting. There were concerns around how to include masters into the node pools without accidentally flooding masters in existing clusters with a bunch of traffic. Some options:
* add master to node pool (with sufficient warning, etc), perhaps add `alpha.service-controller.kubernetes.io/exclude-balancer` automatically where possible. 
* if `externalTrafficPolicy=Local` AND node is master, add it to node pool anyways because user must have intentionally added master toleration (suggested by @justinsb)

Second option seems like a reasonable step forward to allow traffic into the masters without unexpectedly allowing traffic into master nodes. Thoughts?

**Comment by hogepodge** [comment]
@smarterclayton @jhorwit2 has any progress been made on some design work related to this issue?

**Comment by ljani** [comment]
Personally I've lost track of the discussion and I've also switched environments, so I cannot comment whenever or not this is still a problem.

**Comment by pravinsinghal** [comment]
Has this issue been addressed - I am running into a scenario where the node is marked master and the externalTrafficPolicy is set to Local - and the node is getting excluded from master ..

**Comment by andrewsykim** [comment]
This issue has not been resolved yet. I personally would like to see the `SerivceNodeExclusion` feature  (https://github.com/kubernetes/enhancements/issues/724) promoted to beta before doing this though 

/remove-lifecycle stale

**Comment by pravinsinghal** [comment]
What would be the recommended way - if I need a service on node labelled master needs to be exposed  via a LB.  Is there a way I can enable this alpha feature ? From the code it seems the master node role is checked before the service node exclusion ..

**Comment by andrewsykim** [comment]
As of today, you can't add control plane to LBs. Adding masters can lead to unexpected behaviours for users. This definitely warrants more discussion.

**Comment by pravinsinghal** [comment]
This is a custom deployment of k8s stack - where depending on the scenario - its reduces to a single node all in one. Any issues top of the mind that I can run into if I modify local k8s stack to enable it. I Need to unblock.

**Comment by andrewsykim** [comment]
One alternative is to remove the master node role label on the node object but that may have other implications for your cluster.

**Comment by jhorwit2** [comment]
As long as the service isn't using local for the traffic policy then it'll work today. That's because node ports will balance the traffic on workers to the masters when the traffic policy is Cluster (default).

**Comment by jiatongw** [comment]
I'll pick up this soon @andrewsykim

**Comment by andrewsykim** [comment]
Thanks @jiatongw! We'll need a KEP as per https://github.com/kubernetes/enhancements/issues/724 and there's more context from SIG network here https://groups.google.com/forum/#!topic/kubernetes-sig-network/erULVdDkLNw

**Comment by athenabot** [comment]
@jhorwit2 @thockin
If this issue has been triaged, please comment `/remove-triage unresolved`.

If you aren't able to handle this issue, consider unassigning yourself and/or adding the `help-wanted` label.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by andrewsykim** [comment]
/remove-lifecycle stale

This is going to be fixed with https://github.com/kubernetes/enhancements/pull/1083

**Comment by wking** [comment]
This was addressed via kubernetes/enhancements#1144 (see [here][1]) and #90126.

[1]: https://github.com/kubernetes/enhancements/pull/1144/files#diff-55982acee19ca4466736d3b3c150218aR146
